usim/_basics/pipe.py

```diff
@@ -0,0 +1,97 @@
+from typing import AsyncIterable, Optional, Dict
+
+from .._primitives.notification import Notification, suspend
+from .._core.loop import __LOOP_STATE__
+
+
+class Pipe:
+    """
+    .. code:: python3
+
+        network = Pipe(throughput=100)
+
+        async with network.block(throughput=50):
+            await network.transfer(10000, throughput=100)
+    """
+    def __init__(self, throughput: float):
+        assert throughput > 0
+        self.throughput = throughput
+        self._congested = Notification()
+        self._delay_scale = 1.0
+        self._subscriptions: Dict[object, float] = {}
+
+    async def transfer(
+            self, total: float, throughput: Optional[float] = None
+    ):
+        """
+        Wait until some total volume has been transferred
+
+        .. code:: python3
+
+            await network.transfer(total=50e9)
+
+        :param total: absolute volume to transfer before resuming
+        :param throughput: maximum throughput
+        :return:
+        """
+        transferred = 0
```

Comment:
I think it does make sense to ensure that the transfer is non-negative. What do you think?

```suggestion
        assert total > 0
        transferred = 0
```

Reply:
I've added asserts for all inputs now.